source/common/logger/ur_sinks.hpp

```diff
@@ -24,12 +24,15 @@ class Sink {
   public:
     template <typename... Args>
     void log(logger::Level level, const char *fmt, Args &&...args) {
+        std::stringstream buffer;
         if (!skip_prefix) {
-            *ostream << "<" << logger_name << ">";
-            *ostream << "[" << level_to_str(level) << "]: ";
+            buffer << "<" << logger_name << ">";
+            buffer << "[" << level_to_str(level) << "]: ";
         }
-        format(fmt, std::forward<Args &&>(args)...);
-        *ostream << "\n";
+        buffer << format(fmt, std::forward<Args &&>(args)...);
+        buffer << "\n";
+
+        *ostream << buffer.str();
```

Comment:
you are now doing two allocations, one at each step. I think it would be better to have it all reside in the format function so that we avoid one.

Reply:
Restructured a bit.

Reply:
I'm not sure if this is any better. I thought we'd be able to `format("{} ... - ...", actually assemble the prefix here, rest of the args);`. But now that I look at it, I'm not sure if that's possible.
I think we could make the format function optionally take a `ostringstream` instance, and do it like this:
```
std::ostringstream oss;

if (!skip_prefix) {
  oss << "<" << logger_name << ">" << "[" << level_to_str(level) << "]: ";
}

format_oss(oss, fmt, std::forward<Args>(args)...);
{
  scoped_lock();
  *ostream << buffer;
  ...
}
```

Reply:
Added an ostringstream as a required argument (as format() functions are not used in any other place anyway) so that there is only one object used as a buffer.